ykcapi/src/lib.rs

```diff
@@ -61,4 +66,52 @@ mod c_testing {
         Box::leak(bm);
         ret
     }
+
+    #[no_mangle]
+    pub extern "C" fn __yktrace_start_tracing(kind: usize) -> *mut ThreadTracer {
+        let kind: TracingKind = match kind {
+            SW_TRACING => TracingKind::SoftwareTracing,
+            HW_TRACING => TracingKind::HardwareTracing,
+            _ => panic!(),
+        };
+        Box::into_raw(Box::new(start_tracing(kind)))
+    }
+
+    #[no_mangle]
+    pub extern "C" fn __yktrace_stop_tracing(tt: *mut ThreadTracer) -> *mut IRTrace {
+        let tt = unsafe { Box::from_raw(tt) };
+        Box::into_raw(Box::new(tt.stop_tracing().unwrap())) as *mut _
+    }
+
+    #[no_mangle]
+    pub extern "C" fn __yktrace_irtrace_len(trace: *mut IRTrace) -> usize {
+        let trace = unsafe { Box::from_raw(trace) };
+        let ret = trace.len();
+        Box::leak(trace);
+        ret
+    }
+
+    /// Fetches the function name (`res_func`) and the block index (`res_bb`) at position `idx` in
+    /// `trace`. The caller must free() the function name.
+    #[no_mangle]
+    pub extern "C" fn __yktrace_irtrace_get(
+        trace: *mut IRTrace,
+        idx: usize,
+        res_func: *mut *mut c_char,
+        res_bb: *mut usize,
+    ) {
+        let trace = unsafe { Box::from_raw(trace) };
+        let blk = trace.get(idx).unwrap();
+        let c_func = CString::new(blk.func_name()).unwrap();
+        unsafe {
+            *res_func = c_func.into_raw();
+            *res_bb = blk.bb();
+        }
+        Box::leak(trace);
```

Comment:
I think this might better be `mem::forget(trace)`? Basically you just don't want it to be `drop`ed, right?

Reply:
You sir, are correct.

Bjorn had a more concise way of getting a reference without any Box dances (thus effectively borrowing the content), but I forgot how it was done :P

`mem::forget` is fine.

Reply:
Fixed in 2a1af1a2

Reply:
You can convert the `*mut` into a `&` with `unsafe { &*trace }` and that won't be dropped (it's just a reference).